test/jdk/java/net/httpclient/HttpGetInCancelledFuture.java

```diff
@@ -0,0 +1,343 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpClient.Version;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import jdk.internal.net.http.common.OperationTrackers.Tracker;
+import jdk.test.lib.net.SimpleSSLContext;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/*
+ * @test
+ * @bug 8316580
+ * @library /test/lib
+ * @run junit/othervm -Djdk.tracePinnedThreads=full
+ *                   -DuseReferenceTracker=false
+ *                   HttpGetInCancelledFuture
+ * @run junit/othervm -Djdk.tracePinnedThreads=full
+ *                   -DuseReferenceTracker=true
+ *                   HttpGetInCancelledFuture
+ * @summary This test verifies that cancelling a future that
+ * does an HTTP request using the HttpClient doesn't cause
+ * HttpClient::close to block forever.
+ */
+public class HttpGetInCancelledFuture {
+
+    static final boolean useTracker = Boolean.getBoolean("useReferenceTracker");
+
+    static final class TestException extends RuntimeException {
+        public TestException(String message, Throwable cause) {
+            super(message, cause);
+        }
+    }
+
+    static ReferenceTracker TRACKER = ReferenceTracker.INSTANCE;
+
+    HttpClient makeClient(URI uri, Version version, Executor executor) {
+        var builder = HttpClient.newBuilder();
+        if (uri.getScheme().equalsIgnoreCase("https")) {
+            try {
+                builder.sslContext(new SimpleSSLContext().get());
+            } catch (IOException io) {
+                throw new UncheckedIOException(io);
+            }
+        }
+        return builder.connectTimeout(Duration.ofSeconds(1))
+                .executor(executor)
+                .version(version)
+                .build();
+    }
+
+    static final String HOST = "localhost:62057";
```

Comment:
Should we have something listening on that port? Some of our tests used to fail after receiving an unexpected connenction, and 62057 is in the ephemeral range.

Reply:
Good question. I guess it doesn't matter as long as it doesn't reply with a valid response, but if the port is a server port used by some other test running on the machine that only expect a given type/number/sequence of requests it might make the other test fail. Let me experiment with a ServerSocket that doesn't answer, then we might get either TimeoutException or Connection Refused when the backlog is full. That should work as well to test the fix.

Reply:
@djelinski Now done.

Reply:
I'm the submitter of the bug report. I found the issue when the server was down and nothing was listening in the port.

The bug also happens if there is a server listening in the port, the server sends a response and the code running in the subtask throws an Exception. For this case it seems I need to use more concurrent threads to achieve the same effect. For no servers 70 was enough but for a running server I need 200.

The test program I wrote is available at: https://github.com/mikko-apo/jvm21-demo/blob/main/jvm21-demo/src/main/java/fi/iki/apo/HttpGetBurst.java and I use it against helidon/quickstart-standalone-mp-4m2 which is in the same repository.

Reply:
Thanks @mikko-apo - the JBS issue also has a reproducer and I have verified that it no longer fails with the proposed fix.

Reply:
@dfuch Great that you managed to get that fixed so quickly :)